### ISSUE TITLE
Flux uses registry credentials from other namespaces

### ISSUE THREAD

**Issue by bootc** [fluxcd/flux#2723]
**Describe the bug**
Flux appears to be using credentials from other namespaces when accessing our private registry, and is thus failing to check for image updates.

The logs below show the relevant (and slightly redacted) lines for one of our environments with 5 applications in 5 different namespaces, all pulling from the same registry and with the same Secret name. `app1:Secret/gitlab-registry` can only see the `registry.example.com/tcl/app1` tags, and so on. For some reason Flux is picking one of our apps (`app5` in the example below) and using that credential to query the tags on all the images including `app1` through `app4`. In effect the credential is bleeding across namespaces.

**To Reproduce**
Steps to reproduce the behaviour:
0. Create two namespaces with two different Docker registry secrets, with the same name and same registry host.
1. Deploy something to give Flux some images to look for.
2. Notice that Flux is making queries to the registry with only one set of credentials for both namespaces.

**Expected behavior**
It should use the credential only from the relevant namespace.

**Logs**
```
ts=2020-01-06T11:12:05.871370995Z caller=warming.go:180 component=warmer canonical_name=registry.example.com/tcl/app1 auth="{map[registry.example.com:<registry creds for gitlab+deploy-token-12@registry.example.com, from app5:secret/gitlab-registry>]}" err="requesting tags: errors:\ndenied: requested access to the resource is denied\nunauthorized: authentication required\n"
ts=2020-01-06T11:12:21.692306974Z caller=warming.go:180 component=warmer canonical_name=registry.example.com/tcl/app2 auth="{map[registry.example.com:<registry creds for gitlab+deploy-token-12@registry.example.com, from app5:secret/gitlab-registry>]}" err="requesting tags: errors:\ndenied: requested access to the resource is denied\nunauthorized: authentication required\n"
ts=2020-01-06T11:12:31.457550778Z caller=warming.go:180 component=warmer canonical_name=registry.example.com/tcl/app3 auth="{map[registry.example.com:<registry creds for gitlab+deploy-token-12@registry.example.com, from app5:secret/gitlab-registry>]}" err="requesting tags: errors:\ndenied: requested access to the resource is denied\nunauthorized: authentication required\n"
ts=2020-01-06T11:12:42.844377867Z caller=warming.go:180 component=warmer canonical_name=registry.example.com/tcl/app4 auth="{map[registry.example.com:<registry creds for gitlab+deploy-token-12@registry.example.com, from app5:secret/gitlab-registry>]}" err="requesting tags: errors:\ndenied: requested access to the resource is denied\nunauthorized: authentication required\n"
```

**Additional context**
Add any other context about the problem here, e.g 
- Flux version: 1.17.0 (also seen with 1.16.0, but 1.15.0 is fine)
- Helm Operator version: [not relevant]
- Kubernetes version: 1.14 (via EKS) and 1.15 (via RKE)
- Git provider: GitLab
- Container registry provider: GitLab

**Comment by 2opremio** [comment]
I think that the problem was introduced in https://github.com/fluxcd/flux/pull/2520

I am looking into it.

**Comment by 2opremio** [comment]
BTW, thanks a lot for the detailed bug report and for pinpointing in what version the bug was introduced.

**Comment by 2opremio** [comment]
@bootc I have just created a fix for this. See #2728 . It will be included in the next release.

**Comment by bootc** [comment]
Thanks @2opremio! Is there (or will there be) an image of Flux I can test with before it's released?

**Comment by 2opremio** [comment]
I can create one for you. Give me a few minutes

**Comment by 2opremio** [comment]
You can use `2opremio/flux:scope-imagepullsecrets-correctly-a802915a` to test that it works

**Comment by bootc** [comment]
Yes, that looks like it fixes the issue for me. Many thanks! :+1:

**Comment by 2opremio** [comment]
@bootc The fix for this issue has been released with 1.17.1
